File: app/src/main/java/hsk/practice/myvoca/widget/VocaWidget.java
package hsk.practice.myvoca.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.util.Log;
import android.widget.RemoteViews;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;

import java.util.concurrent.Callable;

import Database.Vocabulary;
import Database.source.VocaRepository;
import hsk.practice.myvoca.AppHelper;
import hsk.practice.myvoca.R;

public class VocaWidget extends AppWidgetProvider {

    private VocaRepository repository = VocaRepository.getInstance();
    private AppWidgetManager manager;

    private RemoteViews remoteView;

    private AsyncTask<Void, Void, LiveData<Vocabulary>> showVocaTask;
    private Callable myTask;

    private UpdateWidgetReceiver receiver;
    private int UPDATE_WIDGET_ID = 17;
    public static final String UPDATE_WIDGET = "action.updatewidget.update";

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d("HSK APP", "VocaWidget onReceive(): " + intent.getAction());
        init(context);
        super.onReceive(context, intent);

        if (intent.getAction().equalsIgnoreCase(AppWidgetManager.ACTION_APPWIDGET_UPDATE)) {
            int widgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, -1);
            int[] widgetIds;
            if (widgetId != -1) {
                Log.d("HSK APP", "Widget here");
                widgetIds = new int[]{widgetId};
            } else {
                Log.d("HSK APP", "Widget here 2");
                ComponentName componentName = new ComponentName(context, VocaWidget.class);
                widgetIds = manager.getAppWidgetIds(componentName);
            }
            this.onUpdate(context, manager, widgetIds);
            showRandomVocabulary(context, widgetIds);
        }
    }

    @Override
    public void onEnabled(Context context) {
        Log.d("HSK APP", "VocaWidget onEnabled()");
        super.onEnabled(context);
        init(context);

        AppHelper.loadInstance(context);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        Log.d("HSK APP", "VocaWidget onUpdate()");
        AppHelper.loadInstance(context.getApplicationContext());
        String temp = "";
        for (int id : appWidgetIds) {
            temp += id + " ";
        }
        Log.d("HSK APP", "Widget ids in onUpdate(): " + temp);

        ComponentName componentName = new ComponentName(context, VocaWidget.class);
        RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.widget_layout);
        setPendingIntent(context, remoteViews, appWidgetIds, componentName);
     /*   Intent intent = new Intent(context, VocaWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[0]);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context,5, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        remoteViews.setOnClickPendingIntent(R.id.widget_voca_button, pendingIntent);

        appWidgetManager.updateAppWidget(appWidgetIds, remoteViews);*/
    }

    private void init(Context context) {
        if (remoteView == null) {
            remoteView = new RemoteViews(context.getPackageName(), R.layout.widget_layout);
        }
        if (manager == null) {
            manager = AppWidgetManager.getInstance(context);
        }
    }

    private void setPendingIntent(Context context, RemoteViews remoteViews, int[] appWidgetIds, ComponentName componentName) {
        Intent intent = new Intent(context, VocaWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[0]);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context.getApplicationContext(), appWidgetIds[0], intent, PendingIntent.FLAG_UPDATE_CURRENT);
        remoteViews.setOnClickPendingIntent(R.id.widget_voca_button, pendingIntent);

        manager.updateAppWidget(componentName, remoteViews);

        manager.updateAppWidget(appWidgetIds, remoteViews);
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        Log.d("HSK APP", "Widget onDeleted()");
        super.onDeleted(context, appWidgetIds);
    }

    private void showRandomVocabulary(final Context context, final int[] widgetIds) {
        final LiveData<Vocabulary> randomVocabulary = VocaRepository.getInstance().getRandomVocabulary();
        randomVocabulary.observeForever(new Observer<Vocabulary>() {
            @Override
            public void onChanged(Vocabulary vocabulary) {
                Log.d("HSK APP", "show: " + vocabulary.eng);
                RemoteViews remoteView = new RemoteViews(context.getPackageName(), R.layout.widget_layout);
                remoteView.setTextViewText(R.id.widget_eng, vocabulary.eng);
                remoteView.setTextViewText(R.id.widget_kor, vocabulary.kor);
                ComponentName componentName = new ComponentName(context, VocaWidget.class);
                setPendingIntent(context, remoteView, widgetIds, componentName);

                String temp = "";
                for (int id : widgetIds) {
                    temp += id + " ";
                }
                Log.d("HSK APP", "Widget ids in showRandomVoca(): " + temp);
                randomVocabulary.removeObserver(this);
            }
        });
    }

    public class UpdateWidgetReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            Log.d("HSK APP", "WidgetReceiver onReceive()");
            if (intent.getAction().equalsIgnoreCase(UPDATE_WIDGET)) {
                ComponentName widgetName = new ComponentName(context.getPackageName(), VocaWidget.class.getName());
                int[] widgetId = manager.getAppWidgetIds(widgetName);
                showRandomVocabulary(context, widgetId);
            }
        }
    }
}

File: app/src/main/java/hsk/practice/myvoca/widget/VocaWidget.kt
package hsk.practice.myvoca.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.AsyncTask
import android.util.Log
import android.widget.RemoteViews
import androidx.lifecycle.LiveData
import com.hsk.data.VocaRepository
import com.orhanobut.logger.Logger
import hsk.practice.myvoca.AppHelper
import hsk.practice.myvoca.R
import hsk.practice.myvoca.framework.RoomVocabulary
import hsk.practice.myvoca.framework.VocaPersistenceDatabase
import hsk.practice.myvoca.framework.toRoomVocabulary
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.util.concurrent.Callable

class VocaWidget : AppWidgetProvider() {

    private var manager: AppWidgetManager? = null
    private var remoteView: RemoteViews? = null
    private val showVocaTask: AsyncTask<Void?, Void?, LiveData<RoomVocabulary?>?>? = null
    private val myTask: Callable<*>? = null
    private val receiver: UpdateWidgetReceiver? = null
    private val UPDATE_WIDGET_ID = 17

    private var vocaRepository: VocaRepository? = null

    private val job = SupervisorJob()
    val coroutineScope = CoroutineScope(Dispatchers.Main + job)

    companion object {
        val UPDATE_WIDGET: String = "action.updatewidget.update"
    }

    override fun onReceive(context: Context?, intent: Intent?) {
        if (intent != null) {
            Logger.d("VocaWidget onReceive(): ${intent.action}")
        }
        init(context)
        super.onReceive(context, intent)

        if (intent?.action.equals(AppWidgetManager.ACTION_APPWIDGET_UPDATE, ignoreCase = true)) {
            val widgetId = intent?.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, -1) ?: 0
            val widgetIds: IntArray? = if (widgetId != -1) {
                intArrayOf(widgetId)
            } else {
                val componentName = ComponentName(context!!, VocaWidget::class.java)
                manager?.getAppWidgetIds(componentName)
            }
            onUpdate(context, manager, widgetIds)
            showRandomVocabulary(context, widgetIds)
        }
    }

    override fun onEnabled(context: Context?) {
        Logger.d("VocaWidget onEnabled()")
        super.onEnabled(context)
        init(context)
        AppHelper.loadInstance(context!!)
    }

    override fun onUpdate(context: Context?, appWidgetManager: AppWidgetManager?, appWidgetIds: IntArray?) {
        Logger.d("VocaWidget onUpdate()")
        AppHelper.loadInstance(context!!.applicationContext)
        var temp = ""
        if (appWidgetIds != null) {
            for (id in appWidgetIds) {
                temp += "$id "
            }
        }
        Logger.d("Widget ids in onUpdate(): $temp")
        val componentName = ComponentName(context, VocaWidget::class.java)
        val remoteViews = RemoteViews(context.packageName, R.layout.widget_layout)
        setPendingIntent(context, remoteViews, appWidgetIds, componentName)
        /*   Intent intent = new Intent(context, VocaWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds[0]);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(context,5, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        remoteViews.setOnClickPendingIntent(R.id.widget_voca_button, pendingIntent);

        appWidgetManager.updateAppWidget(appWidgetIds, remoteViews);*/
    }

    private fun init(context: Context?) {
        if (remoteView == null) {
            remoteView = RemoteViews(context!!.packageName, R.layout.widget_layout)
        }
        if (manager == null) {
            manager = AppWidgetManager.getInstance(context)
        }
        if (vocaRepository == null) {
            vocaRepository = VocaRepository(VocaPersistenceDatabase.getInstance(context!!))
        }
    }

    private fun setPendingIntent(context: Context?, remoteViews: RemoteViews?, appWidgetIds: IntArray?, componentName: ComponentName?) {
        val intent = Intent(context, VocaWidget::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetIds?.get(0))
        val pendingIntent = PendingIntent.getBroadcast(context?.applicationContext, appWidgetIds?.get(0)
                ?: -1, intent, PendingIntent.FLAG_UPDATE_CURRENT)
        remoteViews?.setOnClickPendingIntent(R.id.widget_voca_button, pendingIntent)
        manager?.updateAppWidget(componentName, remoteViews)
        manager?.updateAppWidget(appWidgetIds, remoteViews)
    }

    override fun onDeleted(context: Context?, appWidgetIds: IntArray?) {
        Logger.d("Widget onDeleted()")
        super.onDeleted(context, appWidgetIds)
    }

    private fun showRandomVocabulary(context: Context?, widgetIds: IntArray?) = coroutineScope.launch {
        vocaRepository?.getRandomVocabulary()?.toRoomVocabulary()?.let {
            Log.d(AppHelper.LOG_TAG, "show: ${it.eng}")

            val remoteView = RemoteViews(context?.packageName, R.layout.widget_layout)
            remoteView.setTextViewText(R.id.widget_eng, it.eng)
            remoteView.setTextViewText(R.id.widget_kor, it.kor)
            val componentName = ComponentName(context!!, VocaWidget::class.java)
            setPendingIntent(context, remoteView, widgetIds, componentName)

            val temp = widgetIds?.joinToString() ?: ""
            Logger.d("Widget ids in showRandomVoca(): $temp")
        }
    }

    inner class UpdateWidgetReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            Logger.d("WidgetReceiver onReceive()")
            if (intent?.action.equals(UPDATE_WIDGET, ignoreCase = true)) {
                val widgetName = ComponentName(context!!.packageName, VocaWidget::class.java.name)
                val widgetId = manager?.getAppWidgetIds(widgetName)
                showRandomVocabulary(context, widgetId)
            }
        }
    }

}